lua/delaytrain/init.lua

```diff
@@ -77,16 +77,43 @@ end
 function M.enable()
     is_enabled = true
 
+    -- Get an array of all the keys we want to delay regardless of remap status
+    local function get_keypress_array (key, mode_array)
+        local keypress_array = {}
+        for _, mode in ipairs(mode_array) do
+            local keypress = key
+            local remapped = vim.fn.maparg(key, mode, false, true).rhs
+            if remapped then
+                keypress = remapped
+            end
+
+            -- If keypress values differ across modes, add the new value here
+            if keypress_array[keypress] then
+                table.insert(keypress_array[keypress], mode)
+            -- Append any mode that has an equal value to previously seen modes
+            else
+                keypress_array[keypress] = {mode}
+            end
+        end
+
+      -- If the keypress value is equal across modes, the array will only have one value,
+      -- so vim.keymap.set will be called once with a copy of the original mode_array
+      return keypress_array
+    end
+
```

Comment:
Looping through the original mode_array, we generate a unique table of `[keypress] = {modes}` kv pairs, setting `keypress` to a remap value if there is one.

This gets us the appropriate mode array to call with `vim.keymap.set` for each unique keypress value. We use a `for` loop below to call it, but if the value is the same across all modes in the original `mode_array`, the for loop only loops once.

I made a nested function to try to keep the main function neat. I hope this isn't adding more bloat to your nice neat code then you were bargaining for. I made the code comments a bit verbose on purpose for now but I'd like them to be more concise.

This commit makes my comments here https://github.com/ja-ford/delaytrain.nvim/pull/6#discussion_r1032851984 outdated. I still need to look at the issues you mentioned here https://github.com/ja-ford/delaytrain.nvim/pull/6#issuecomment-1327692707

Reply:
Once again sorry for the delay here.

So strange. I was trying to reproduce my issue with item 2 in https://github.com/ja-ford/delaytrain.nvim/pull/6#issuecomment-1327692707 and with your current fix it appears to be working fine. This was either fixed by your changes or my config from a month ago was having issues. Either way I think this is great and I'm nearly ready to push the button. I haven't been able to find a breaking use-case so far. I'm personally a fan of verbose comments but I think your latest changes are a happy medium.

I have one last nitpick and that's about the naming of `get_keypress_array`. It's a bit vague and we're technically using it to find the actual key's mappings. I think naming it `get_key_mappings` and naming the table `key_mappings` makes it slightly more clear that this code is searching for key mappings (even if it turns out the key isn't actually remapped). I'm open to other suggestions here though.

Once we're all set with that feel free to take this out of draft status and I'll merge it. Thanks again for working through this!

Reply:
Renamed the offenders and marked ready for review.

Thanks for guiding me through all the bugs, you're great at explaining what needs to be done and why. I had fun doing this!

Reply:
And thanks for contributing! For a first PR, great job.